lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,192 @@
+			strings.Contains(err.Error(), "no space left on device") {
+			return nil, errors.New("Please increase inotify limits, see http://bit.ly/1PxkdUC for more information.")
+		}
+		return nil, fmt.Errorf(
+			"Failed to install inotify handler for %s. Error: %s",
+			path, err)
+	}
+	l.Debugf("Setup filesystem notification for %s", path)
+	return c, nil
+}
+
+func (watcher *FsWatcher) watchFilesystem() {
+	watcher.notifyTimer = *time.NewTimer(watcher.notifyDelay)
+	defer func() {
+		watcher.notifyTimer.Stop()
+	}()
+	finishedFileEventSubscription := events.Default.Subscribe(events.ItemFinished)
+	for {
+		watcher.resetNotifyTimerIfNeeded()
+		select {
+		case event, _ := <- watcher.fsEventChan:
+			watcher.speedUpNotifyTimer()
+			watcher.storeFsEvent(event)
+		case <-watcher.notifyTimer.C:
+			watcher.sendStoredEventsToModelOrSlowDownTimer()
+		case event := <-finishedFileEventSubscription.C():
+			watcher.skipPathChangedByUs(event)
+		}
+	}
+}
+
+func (watcher *FsWatcher) newFsEvent(eventPath string) *FsEvent {
+	if isSubpath(eventPath, watcher.folderPath) {
+		path := relativePath(eventPath, watcher.folderPath)
+		if !isSpecialPath(path) {
+			return &FsEvent{path}
+		}
+	}
+	return nil
+}
+
+func relativePath(path string, folderPath string) string {
+	subPath, _ := filepath.Rel(folderPath, path)
+	return subPath
+}
+
+func isSubpath(path string, folderPath string) bool {
+	if len(path) > 1 && os.IsPathSeparator(path[len(path) - 1]) {
+		path = path[0:len(path)-1]
+	}
+	if len(folderPath) > 1 && os.IsPathSeparator(folderPath[len(folderPath) - 1]) {
+		folderPath = folderPath[0:len(folderPath)-1]
+	}
+	return strings.HasPrefix(path, folderPath)
+}
+
+func (watcher *FsWatcher) resetNotifyTimerIfNeeded() {
+	if watcher.notifyTimerNeedsReset {
+	l.Debugf("Resetting notifyTimer to %#v\n", watcher.notifyDelay)
+		watcher.notifyTimer.Reset(watcher.notifyDelay)
+		watcher.notifyTimerNeedsReset = false
+	}
+}
+
+func (watcher *FsWatcher) speedUpNotifyTimer() {
+	if watcher.notifyDelay != fastNotifyDelay {
+		watcher.notifyDelay = fastNotifyDelay
+		l.Debugf("Speeding up notifyTimer to %#v\n", watcher.notifyDelay)
+		watcher.notifyTimerNeedsReset = true
+	}
+}
+
+func (watcher *FsWatcher) slowDownNotifyTimer() {
+	if watcher.notifyDelay != slowNotifyDelay {
+		watcher.notifyDelay = slowNotifyDelay
+		l.Debugf("Slowing down notifyTimer to %#v\n", watcher.notifyDelay)
+		watcher.notifyTimerNeedsReset = true
+	}
+}
+
+const (
+	slowNotifyDelay = time.Duration(60) * time.Second
+	fastNotifyDelay = time.Duration(500) * time.Millisecond
+)
+
+func (watcher *FsWatcher) addEvent(event FsEvent) {
+	watcher.fsEvents[event.path] = event
+}
+
+func (watcher *FsWatcher) removeEventIfPresent(path string) {
+	delete(watcher.fsEvents, path)
+}
+
+func (watcher *FsWatcher) storeFsEvent(event notify.EventInfo) {
+	newEvent := watcher.newFsEvent(event.Path())
+	if newEvent != nil {
+		watcher.addEvent(*newEvent)
```

Comment:
I usually start by writing functions with informative names and optimize later. This has been changed already.